main.go

```diff
@@ -52,52 +51,62 @@ func main() {
 	}
 }
 
-func master(stopChan *chan struct{}) {
+func master(stopChan chan<- string) {
 	var err error
 	for {
 		// has master changed from last time?
 		masterAddr, err = getMasterAddr(saddr, *masterName)
 		if err != nil {
-			log.Println(err)
+			log.Println("[MASTER] Error polling for new master: %s", err)
 		}
-		if masterAddr.String() != prevMasterAddr.String() {
-			fmt.Printf("Master Address changed. Closing stopChan. %s v. %s \n", masterAddr.String(), prevMasterAddr.String())
-			close(*stopChan)
-			*stopChan = make(chan struct{})
+		if err == nil && masterAddr.String() != prevMasterAddr.String() {
+			log.Printf("[MASTER] Master Address changed from %s to %s \n", prevMasterAddr.String(), masterAddr.String())
+			prevMasterAddr = masterAddr
+			stopChan <- masterAddr.String()
 		}
-		prevMasterAddr = masterAddr
-		time.Sleep(1 * time.Second)
+		time.Sleep(500 * time.Second)
 	}
 }
 
-func pipe(r io.Reader, w io.WriteCloser) {
-	io.Copy(w, r)
-	w.Close()
-	fmt.Println("Closing pipe")
+func pipe(r net.Conn, w net.Conn, proxyChan chan<- string) {
+	bytes, err := io.Copy(w, r)
+	log.Println("[PROXY %s => %s] Shutting down stream; transferred %s bytes: %s", w.RemoteAddr().String(), r.RemoteAddr().String(), bytes, err)
+	close(proxyChan)
 }
 
 // pass a stopChan to the go routtine
-func proxy(local io.ReadWriteCloser, remoteAddr *net.TCPAddr, stopChan chan struct{}) {
-	fmt.Printf("Opening a new connection on remoteAddr, %s\n", remoteAddr)
-	remote, err := net.DialTCP("tcp", nil, remoteAddr)
+func proxy(client *net.TCPConn, redisAddr *net.TCPAddr, stopChan <-chan string) {
+	redis, err := net.DialTimeout("tcp4", redisAddr.String(), 50*time.Millisecond)
 	if err != nil {
-		log.Println(err)
-		local.Close()
+		log.Println("[PROXY %s => %s] Can't establish connection: %s", client.RemoteAddr().String(), redisAddr.String(), err)
+		client.Close()
 		return
 	}
-	go pipe(local, remote)
-	go pipe(remote, local)
-	<-stopChan // read from stopChan
-	fmt.Println("Closing Proxy")
-	local.Close()
+
+	log.Printf("[PROXY %s => %s] New connection\n", client.RemoteAddr().String(), redisAddr.String())
+	defer client.Close()
+	defer redis.Close()
+
+	clientChan := make(chan string)
+	redisChan := make(chan string)
+
+	go pipe(client, redis, redisChan)
+	go pipe(redis, client, clientChan)
+
+	select {
+	case <-stopChan:
```

Comment:
How does this work when there are multiple connections? If we have multiple `proxy` goroutines listening to this channel and we push a value into it, will all subscribers receive it? I thought one subscriber would get the value.